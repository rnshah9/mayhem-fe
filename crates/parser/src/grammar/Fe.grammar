file_input: NEWLINE ENDMARKER | module_stmt+ ENDMARKER

module_stmt: import_stmt | type_def | contract_def | func_def

########################### path ######################################

path: NAME ('::' NAME)*

trailing_delim_path: path '::'

########################### use_stmt ##################################

use_stmt: 'use' use_tree NEWLINE

use_tree: (simple_use_tree | glob_use_tree | nested_use_tree)

simple_use_tree: path ['as' NAME]
glob_use_tree: trailing_delim_path '*'
nested_use_tree: trailing_delim_path '{' [use_tree (',' use_tree)* [',']] '}'

########################### type_def #################################

type_def: 'type' NAME '=' type_desc NEWLINE

########################### contract_def #############################

contract_def: 'contract' NAME '{' contract_stmt* '}'
contract_stmt: contract_field | event_def | func_def
contract_field: [contract_field_qual] NAME ':' type_desc NEWLINE

event_def: 'event' NAME '{' event_field* '}'
event_field: [event_field_qual] NAME ':' type_desc NEWLINE

func_def:
    [func_qual] 'fn' NAME '(' [arg_list] ')' ['->' base_type] ':' block
arg_list: arg_def (',' arg_def)* [',']
arg_def: NAME ':' type_desc | 'self'

type_desc: map_type | base_type
map_type:
    # This is required because of the ambiguous right-shift token
    'map' '<' base_type ',' 'map' '<' base_type ',' type_desc '>>' |
    'map' '<' base_type ',' type_desc '>'
base_type: NAME arr_list
arr_list: ('[' NUMBER ']')*

contract_field_qual: 'const' | 'pub'
event_field_qual: 'idx'
func_qual: 'pub'

########################### func_stmt ###############################

func_stmt:  compound_stmt | simple_stmt
simple_stmt: small_stmt (';' small_stmt)* [';']

small_stmt: return_stmt | assert_stmt | emit_stmt | break_stmt |
            continue_stmt | revert_stmt | vardecl_stmt | assign_stmt | augassign_stmt |
            exprs

return_stmt: 'return' [exprs]
assert_stmt: 'assert' expr [',' expr]
emit_stmt: 'emit' expr

break_stmt: 'break'
continue_stmt: 'continue'
revert_stmt: 'revert'

vardecl_stmt: target ':' type_desc ['=' expr]
assign_stmt: (targets '=')+ exprs
augassign_stmt: target augassign exprs

augassign: '+=' | '-=' | '*=' | '/=' | '%=' | '&=' | '|=' | '^='
           | '<<=' | '>>=' | '**=' | '//='

compound_stmt: if_stmt | while_stmt | for_stmt

if_stmt: 'if' expr block [else_block]
else_block: 'else'  block

while_stmt: 'while' expr block [else_block]

for_stmt: 'for' targets 'in' exprs block [else_block]

block: '{' (simple_stmt | func_stmt+) '}'

exprs: expr (',' expr)* [',']

expr: disjunct ['if' disjunct 'else' expr]
disjunct: conjunct ('or' conjunct)*
conjunct: comparison ('and' comparison)*
comparison: 'not'* bitwise_or (comp_op bitwise_or)*
comp_op: '<' | '<=' | '==' | '>=' | '>' | '!=' | ['not'] 'in' | 'is' ['not']

bitwise_or: bitwise_xor ('|' bitwise_xor)*
bitwise_xor: bitwise_and ('^' bitwise_and)*
bitwise_and: shift_expr ('&' shift_expr)*
shift_expr: sum (('<<'|'>>') sum)*

sum: term (('+' term | '-' term))*
term: factor (('*' factor | '/' factor | '//' factor | '%' factor))*
factor: ('+' | '-' | '~') factor | power
power: primary '**' factor | primary
primary: atom (attr_tail | index_tail | call_tail)*

slices: slice (',' slice)* [',']
slice: [expr] ':' [expr] [':' [expr]] | expr

atom: list | listcomp | group | tuple | NAME | STRING+ | NUMBER | '...'
list: '[' [exprs] ']'
listcomp: '[' expr for_if_clauses ']'
tuple: '(' [exprs] ')'
group: '(' expr ')'
for_if_clauses: ('for' targets 'in' expr ('if' expr)*)+

args: kwargs | expr [',' args]
kwargs: kwarg (',' kwarg)*
kwarg: NAME '=' expr

targets: target (',' target)* [',']
target: atom t_tail+ | t_atom t_tail*
t_atom: NAME | '(' [targets] ')' | '[' [targets] ']'

t_tail: call_tail* (attr_tail | index_tail)

attr_tail: '.' NAME
index_tail: '[' slices ']'
call_tail: '(' [args] ')'
