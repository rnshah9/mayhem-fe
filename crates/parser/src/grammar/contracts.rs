use super::functions::parse_fn_def;
use super::types::{parse_event_def, parse_field, parse_opt_qualifier};

use crate::ast::{Contract, ContractStmt};
use crate::node::{Node, Span};
use crate::{ParseFailed, ParseResult, Parser, TokenKind};

// Rule: all "statement" level parse functions consume their trailing
// newline(s), either directly or via a function they call.
// This is required to parse an `if` block, because we need to peek past the
// trailing newlines to check whether it's followed by an `else` block, and is
// done for all statements for consistency.

/// Parse a contract definition.
/// # Panics
/// Panics if the next token isn't `contract`.
pub fn parse_contract_def(
    par: &mut Parser,
    contract_pub_qual: Option<Span>,
) -> ParseResult<Node<Contract>> {
    let contract_tok = par.assert(TokenKind::Contract);
    let contract_name = par.expect_with_notes(
        TokenKind::Name,
        "failed to parse contract definition",
        |_| vec!["Note: `contract` must be followed by a name, which must start with a letter and contain only letters, numbers, or underscores".into()],
    )?;

    let mut span = contract_tok.span + contract_name.span;
    par.enter_block(span, "contract definition")?;

    let mut fields = vec![];
    let mut defs = vec![];

    loop {
        par.eat_newlines();
        let mut pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
        let const_qual = parse_opt_qualifier(par, TokenKind::Const);
        if pub_qual.is_none() && const_qual.is_some() && par.peek() == Some(TokenKind::Pub) {
            pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
            par.error(
                pub_qual.unwrap() + const_qual,
                "`const pub` should be written `pub const`",
            );
        }

        match par.peek_or_err()? {
            TokenKind::Name => {
                let field = parse_field(par, pub_qual, const_qual)?;
                if !defs.is_empty() {
                    par.error(field.span, "contract field definitions must come before any function or event definitions");
                }
                fields.push(field);
            }
            TokenKind::Fn | TokenKind::Unsafe => {
                if let Some(span) = const_qual {
                    par.error(
                        span,
                        "`const` qualifier can't be used with function definitions",
                    );
                }
                defs.push(ContractStmt::Function(parse_fn_def(par, pub_qual)?));
            }
            TokenKind::Event => {
                if let Some(span) = pub_qual {
                    par.error(
                        span,
                        "`pub` qualifier can't be used with contract-level event definitions",
                    );
                }
                if let Some(span) = const_qual {
                    par.error(
                        span,
                        "`const` qualifier can't be used with event definitions",
                    );
                }
                defs.push(ContractStmt::Event(parse_event_def(par, None)?));
            }
            TokenKind::BraceClose => {
                span += par.next()?.span;
                break;
            }
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    &tok,
                    "failed to parse contract definition body",
                    vec![],
                );
                return Err(ParseFailed);
            }
        };
    }

    Ok(Node::new(
        Contract {
            name: Node::new(contract_name.text.into(), contract_name.span),
            fields,
            body: defs,
            pub_qual: contract_pub_qual,
        },
        span,
    ))
}
