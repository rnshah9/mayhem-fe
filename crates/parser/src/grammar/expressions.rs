use crate::ast::{self, CallArg, Expr, GenericArg, Path};
use crate::node::Node;
use crate::{Label, ParseFailed, ParseResult, Parser, Token, TokenKind};

use super::types::parse_generic_args;

use if_chain::if_chain;

// Expressions are parsed in Pratt's top-down operator precedence style.
// See this article for a nice explanation:
// <https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html>

/// Parse an expression, starting with the next token.
pub fn parse_expr(par: &mut Parser) -> ParseResult<Node<Expr>> {
    parse_expr_with_min_bp(par, 0)
}

/// Parse an expression, stopping if/when we reach an operator that binds less
/// tightly than given binding power.
pub fn parse_expr_with_min_bp(par: &mut Parser, min_bp: u8) -> ParseResult<Node<Expr>> {
    let mut expr_head = parse_expr_head(par)?;

    while let Some(op) = par.peek() {
        if let Some(lbp) = postfix_binding_power(op) {
            if lbp < min_bp {
                break;
            }

            expr_head = match op {
                TokenKind::ParenOpen => {
                    let args = parse_call_args(par)?;
                    let span = expr_head.span + args.span;
                    Node::new(
                        Expr::Call {
                            func: Box::new(expr_head),
                            generic_args: None,
                            args,
                        },
                        span,
                    )
                }
                TokenKind::BracketOpen => {
                    par.next()?;
                    let index = parse_expr(par)?;
                    let rbracket = par.expect(
                        TokenKind::BracketClose,
                        "failed to parse subscript expression",
                    )?;
                    let span = expr_head.span + rbracket.span;
                    Node::new(
                        Expr::Subscript {
                            value: Box::new(expr_head),
                            index: Box::new(index),
                        },
                        span,
                    )
                }
                TokenKind::If => {
                    par.next()?;
                    let test = parse_expr(par)?;
                    par.expect(
                        TokenKind::Else,
                        "failed to parse ternary `if-else` expression",
                    )?;
                    let else_val = parse_expr(par)?;
                    let span = expr_head.span + else_val.span;
                    Node::new(
                        Expr::Ternary {
                            if_expr: Box::new(expr_head),
                            test: Box::new(test),
                            else_expr: Box::new(else_val),
                        },
                        span,
                    )
                }
                _ => unreachable!(), // patterns above must match those in `postfix_binding_power`
            };
            continue;
        }

        if matches!(op, TokenKind::Lt) {
            let mut bt_par = par.as_bt_parser();
            if_chain! {
                if let Ok(generic_args) = parse_generic_args(&mut bt_par);
                if matches!(bt_par.peek(), Some(TokenKind::ParenOpen));
                if let Ok(args) = parse_call_args(&mut bt_par);
                then {
                    let span = expr_head.span + args.span;
                    expr_head = Node::new(
                        Expr::Call {
                            func: Box::new(expr_head),
                            generic_args: Some(generic_args),
                            args,
                        },
                        span,
                    );
                    bt_par.accept();
                    continue;
                }
            }
        }

        if let Some((lbp, rbp)) = infix_binding_power(op) {
            if lbp < min_bp {
                break;
            }

            let op_tok = par.next()?;
            let rhs = parse_expr_with_min_bp(par, rbp)?;
            expr_head = infix_op(par, expr_head, &op_tok, rhs)?;
            continue;
        }
        break;
    }

    Ok(expr_head)
}

/// Parse call arguments
pub fn parse_call_args(par: &mut Parser) -> ParseResult<Node<Vec<Node<CallArg>>>> {
    use TokenKind::*;
    let lparen = par.assert(ParenOpen);
    let mut args = vec![];
    let mut span = lparen.span;
    loop {
        if par.peek_or_err()? == ParenClose {
            span += par.next()?.span;
            break;
        }

        let arg = parse_expr(par)?;
        match par.peek_or_err()? {
            TokenKind::Eq => {
                let eq = par.next()?;
                if let Expr::Name(name) = arg.kind {
                    par.fancy_error(
                        "Syntax error in argument list",
                        vec![Label::primary(eq.span, "unexpected `=`".to_string())],
                        vec![
                            "Argument labels should be followed by `:`.".to_string(),
                            format!("Hint: try `{}:`", name),
                            "If this is a variable assignment, it must be a standalone statement."
                                .to_string(),
                        ],
                    );
                    let value = parse_expr(par)?;
                    let span = arg.span + value.span;
                    args.push(Node::new(
                        CallArg {
                            label: Some(Node::new(name, arg.span)),
                            value,
                        },
                        span,
                    ));
                } else {
                    par.fancy_error(
                        "Syntax error in argument list",
                        vec![Label::primary(eq.span, "unexpected `=`".to_string())],
                        vec![],
                    );
                }
            }

            TokenKind::Colon => {
                let sep_tok = par.next()?;
                let value = parse_expr(par)?;
                if let Expr::Name(name) = arg.kind {
                    let span = arg.span + value.span;
                    args.push(Node::new(
                        CallArg {
                            label: Some(Node::new(name, arg.span)),
                            value,
                        },
                        span,
                    ));
                } else {
                    par.fancy_error(
                        "Syntax error in function call argument list",
                        vec![
                            Label::primary(
                                sep_tok.span,
                                "In a function call, `:` is used for named arguments".to_string(),
                            ),
                            Label::secondary(
                                arg.span,
                                "this should be a function parameter name".to_string(),
                            ),
                        ],
                        vec![],
                    );
                    return Err(ParseFailed);
                }
            }
            _ => {
                let span = arg.span;
                args.push(Node::new(
                    CallArg {
                        label: None,
                        value: arg,
                    },
                    span,
                ));
            }
        }
        if par.peek_or_err()? == Comma {
            par.next()?;
        } else {
            span += par
                .expect(ParenClose, "failed to parse function call argument list")?
                .span;
            break;
        }
    }

    Ok(Node::new(args, span))
}

/// Try to build an expression starting with the given token.
fn parse_expr_head(par: &mut Parser) -> ParseResult<Node<Expr>> {
    use TokenKind::*;

    match par.peek_or_err()? {
        Name | SelfValue | Int | Hex | Octal | Binary | Text | True | False => {
            let tok = par.next()?;
            Ok(atom(par, &tok))
        }
        Plus | Minus | Not | Tilde => {
            let op = par.next()?;
            let operand = parse_expr_with_min_bp(par, prefix_binding_power(op.kind))?;
            unary_op(par, &op, operand)
        }
        ParenOpen => parse_group_or_tuple(par),
        BracketOpen => parse_list_or_repeat(par),
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                &tok,
                format!("Unexpected token while parsing expression: `{}`", tok.text),
                vec![],
            );
            Err(ParseFailed)
        }
    }
}

/// Specifies how tightly a prefix unary operator binds to its operand.
fn prefix_binding_power(op: TokenKind) -> u8 {
    use TokenKind::*;
    match op {
        Not => 65,
        Plus | Minus | Tilde => 135,
        _ => panic!("Unexpected unary op token: {:?}", op),
    }
}

/// Specifies how tightly does an infix operator bind to its left and right
/// operands. See https://docs.python.org/3/reference/expressions.html#operator-precedence
fn infix_binding_power(op: TokenKind) -> Option<(u8, u8)> {
    use TokenKind::*;

    let bp = match op {
        // assignment expr `:=`?
        // lambda?
        // Comma => (40, 41),
        Or => (50, 51),
        And => (60, 61),
        // prefix Not => 65

        // all comparisons are the same
        Lt | LtEq | Gt | GtEq | NotEq | EqEq => (70, 71),

        Pipe => (80, 81),
        Hat => (90, 91),
        Amper => (100, 101),
        LtLt | GtGt => (110, 111),
        Plus | Minus => (120, 121),
        Star | Slash | Percent => (130, 131),
        // Prefix Plus | Minus | Tilde => 135
        StarStar => (141, 140),
        Dot => (150, 151),
        ColonColon => (160, 161),
        _ => return None,
    };
    Some(bp)
}

/// Specifies how tightly a postfix operator binds to its operand.
/// We don't have any "real" postfix operators (like `?` in rust),
/// but we treat `[`, `(`, and ternary `if` as though they're postfix
/// operators.
fn postfix_binding_power(op: TokenKind) -> Option<u8> {
    use TokenKind::*;
    match op {
        If => Some(35), // ternary
        BracketOpen | ParenOpen => Some(150),
        _ => None,
    }
}

/// Parse a square-bracket list expression, eg. `[1, 2, x]` or `[true; 42]`
fn parse_list_or_repeat(par: &mut Parser) -> ParseResult<Node<Expr>> {
    let lbracket = par.assert(TokenKind::BracketOpen);
    let elts = parse_expr_list(par, &[TokenKind::BracketClose, TokenKind::Semi], None)?;

    if elts.len() == 1 {
        if par.peek() == Some(TokenKind::BracketClose) {
            let rbracket = par.assert(TokenKind::BracketClose);
            let span = lbracket.span + rbracket.span;
            Ok(Node::new(Expr::List { elts }, span))
        } else if par.peek() == Some(TokenKind::Semi) {
            par.assert(TokenKind::Semi);

            let len = if par.peek() == Some(TokenKind::BraceOpen) {
                // handle `{ ... }` const expression
                let brace_open = par.next()?;
                let expr = parse_expr(par)?;
                if !matches!(par.peek(), Some(TokenKind::BraceClose)) {
                    par.error(brace_open.span, "missing closing delimiter `}`");
                    return Err(ParseFailed);
                }
                let brace_close = par.assert(TokenKind::BraceClose);
                let span = brace_open.span + brace_close.span;
                Box::new(Node::new(GenericArg::ConstExpr(expr), span))
            } else {
                // handle const expression without braces
                let expr = parse_expr(par)?;
                let span = expr.span;
                Box::new(Node::new(GenericArg::ConstExpr(expr), span))
            };

            let rbracket = par.assert(TokenKind::BracketClose);
            let span = lbracket.span + rbracket.span;
            Ok(Node::new(
                Expr::Repeat {
                    value: Box::new(elts[0].clone()),
                    len,
                },
                span,
            ))
        } else {
            par.error(lbracket.span, "expected `]` or `;`");
            Err(ParseFailed)
        }
    } else {
        let rbracket = par.assert(TokenKind::BracketClose);
        let span = lbracket.span + rbracket.span;
        Ok(Node::new(Expr::List { elts }, span))
    }
}

/// Parse a paren-wrapped expression, which might turn out to be a tuple
/// if it contains commas.
fn parse_group_or_tuple(par: &mut Parser) -> ParseResult<Node<Expr>> {
    use TokenKind::*;
    let lparen = par.assert(ParenOpen);
    if par.peek_or_err()? == ParenClose {
        let rparen = par.next()?;
        let span = lparen.span + rparen.span;
        return Ok(Node::new(Expr::Unit, span));
    }

    let elem = parse_expr(par)?;
    match par.peek_or_err()? {
        ParenClose => {
            // expr wrapped in parens
            let rparen = par.next()?;
            let span = lparen.span + rparen.span;
            Ok(Node::new(elem.kind, span))
        }
        Comma => {
            // tuple
            par.next()?;
            let elts = parse_expr_list(par, &[ParenClose], Some(elem))?;
            let rparen = par.expect(ParenClose, "failed to parse tuple expression")?;
            let span = lparen.span + rparen.span;
            Ok(Node::new(Expr::Tuple { elts }, span))
        }
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                &tok,
                "Unexpected token while parsing expression in parentheses",
                vec![],
            );
            Err(ParseFailed)
        }
    }
}

/// Parse some number of comma-separated expressions, until `end_marker` is
/// `peek()`ed.
fn parse_expr_list(
    par: &mut Parser,
    end_markers: &[TokenKind],
    head: Option<Node<Expr>>,
) -> ParseResult<Vec<Node<Expr>>> {
    let mut elts = vec![];
    if let Some(elem) = head {
        elts.push(elem);
    }
    loop {
        let next = par.peek_or_err()?;
        if end_markers.contains(&next) {
            break;
        }
        elts.push(parse_expr(par)?);
        match par.peek_or_err()? {
            TokenKind::Comma => {
                par.next()?;
            }
            tk if end_markers.contains(&tk) => break,
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    &tok,
                    "Unexpected token while parsing list of expressions",
                    vec![],
                );
                return Err(ParseFailed);
            }
        }
    }

    Ok(elts)
}

/* node building utils */

/// Create an "atom" expr from the given `Token` (`Name`, `Num`, `Bool`, etc)
fn atom(par: &mut Parser, tok: &Token) -> Node<Expr> {
    use TokenKind::*;

    let expr = match tok.kind {
        Name | SelfValue => Expr::Name(tok.text.into()),
        Int | Hex | Octal | Binary => Expr::Num(tok.text.into()),
        True | False => Expr::Bool(tok.kind == True),
        Text => {
            if let Some(string) = unescape_string(tok.text) {
                Expr::Str(string.into())
            } else {
                par.error(tok.span, "String contains an invalid escape sequence");
                Expr::Str(tok.text.into())
            }
        }
        _ => panic!("Unexpected atom token: {:?}", tok),
    };
    Node::new(expr, tok.span)
}

fn unescape_string(quoted_string: &str) -> Option<String> {
    let inner = &quoted_string[1..quoted_string.len() - 1];
    unescape::unescape(inner)
}

/// Create an expr from the given infix operator and operands.
fn infix_op(
    par: &mut Parser,
    left: Node<Expr>,
    op: &Token,
    right: Node<Expr>,
) -> ParseResult<Node<Expr>> {
    use TokenKind::*;
    let expr = match op.kind {
        Or | And => bool_op(left, op, right),

        Amper | Hat | Pipe | LtLt | GtGt | Plus | Minus | Star | Slash | Percent | StarStar => {
            bin_op(left, op, right)
        }

        Lt | LtEq | Gt | GtEq | NotEq | EqEq => comp_op(left, op, right),

        Dot => {
            let span = left.span + right.span;
            match right.kind {
                Expr::Name(name) => Node::new(
                    Expr::Attribute {
                        value: Box::new(left),
                        attr: Node::new(name, right.span),
                    },
                    span,
                ),
                Expr::Num(_num) => {
                    par.error(span, "floats not supported");
                    return Err(ParseFailed);
                }
                Expr::Call {
                    func,
                    generic_args,
                    args,
                } => {
                    let func_span = left.span + func.span;
                    let func = Box::new(Node::new(
                        Expr::Attribute {
                            value: Box::new(left),
                            attr: {
                                if let Expr::Name(name) = func.kind {
                                    Node::new(name, func.span)
                                } else {
                                    par.fancy_error(
                                        "failed to parse attribute expression",
                                        vec![Label::primary(func.span, "expected a name")],
                                        vec![],
                                    );
                                    return Err(ParseFailed);
                                }
                            },
                        },
                        func_span,
                    ));

                    Node::new(
                        Expr::Call {
                            func,
                            generic_args,
                            args,
                        },
                        span,
                    )
                }
                _ => {
                    par.fancy_error(
                        "failed to parse attribute expression",
                        vec![Label::primary(right.span, "expected a name")],
                        vec![],
                    );
                    return Err(ParseFailed);
                }
            }
        }
        ColonColon => {
            let mut path = match left.kind {
                Expr::Name(name) => Path {
                    segments: vec![Node::new(name, left.span)],
                },
                Expr::Path(path) => path,
                _ => {
                    par.fancy_error(
                        "failed to parse path expression",
                        vec![
                            Label::secondary(op.span, "path delimiter".to_string()),
                            Label::primary(left.span, "expected a name"),
                        ],
                        vec![],
                    );
                    return Err(ParseFailed);
                }
            };

            // `right` can't be a Path (rbp > lbp); only valid option is `Name`
            match right.kind {
                Expr::Name(name) => {
                    path.segments.push(Node::new(name, right.span));
                    Node::new(Expr::Path(path), left.span + right.span)
                }
                _ => {
                    par.fancy_error(
                        "failed to parse path expression",
                        vec![
                            Label::secondary(op.span, "path delimiter".to_string()),
                            Label::primary(right.span, "expected a name"),
                        ],
                        vec![],
                    );
                    return Err(ParseFailed);
                }
            }
        }
        _ => panic!("Unexpected infix op token: {:?}", op),
    };
    Ok(expr)
}

/// Create an `Expr::BoolOperation` node for the given operator and operands.
fn bool_op(left: Node<Expr>, op: &Token, right: Node<Expr>) -> Node<Expr> {
    use TokenKind::*;
    let astop = match op.kind {
        And => ast::BoolOperator::And,
        Or => ast::BoolOperator::Or,
        _ => panic!(),
    };

    let span = left.span + right.span;
    Node::new(
        Expr::BoolOperation {
            left: Box::new(left),
            op: Node::new(astop, op.span),
            right: Box::new(right),
        },
        span,
    )
}

/// Create an `Expr::BinOperation` node for the given operator and operands.
fn bin_op(left: Node<Expr>, op: &Token, right: Node<Expr>) -> Node<Expr> {
    use ast::BinOperator::*;
    use TokenKind::*;

    let astop = match op.kind {
        Amper => BitAnd,
        Hat => BitXor,
        Pipe => BitOr,
        LtLt => LShift,
        GtGt => RShift,
        Plus => Add,
        Minus => Sub,
        Star => Mult,
        Slash => Div,
        Percent => Mod,
        StarStar => Pow,
        _ => panic!(),
    };

    let span = left.span + right.span;
    Node::new(
        Expr::BinOperation {
            left: Box::new(left),
            op: Node::new(astop, op.span),
            right: Box::new(right),
        },
        span,
    )
}

/// Create an `Expr::UnaryOperation` node for the given operator and operands.
fn unary_op(par: &mut Parser, op: &Token, operand: Node<Expr>) -> ParseResult<Node<Expr>> {
    use ast::UnaryOperator;
    use TokenKind::*;

    let astop = match op.kind {
        Tilde => UnaryOperator::Invert,
        Not => UnaryOperator::Not,
        Minus => UnaryOperator::USub,
        Plus => {
            par.fancy_error(
                "unary plus not supported",
                vec![Label::primary(op.span, "consider removing the '+'")],
                vec![],
            );
            return Err(ParseFailed);
        }
        _ => panic!(),
    };

    let span = op.span + operand.span;
    Ok(Node::new(
        Expr::UnaryOperation {
            op: Node::new(astop, op.span),
            operand: Box::new(operand),
        },
        span,
    ))
}

/// Create an `Expr::CompOperation` node for the given operator and operands.
fn comp_op(left: Node<Expr>, op: &Token, right: Node<Expr>) -> Node<Expr> {
    use ast::CompOperator;
    use TokenKind::*;
    let astop = match op.kind {
        In => todo!("in"), // CompOperator::In,
        Lt => CompOperator::Lt,
        LtEq => CompOperator::LtE,
        Gt => CompOperator::Gt,
        GtEq => CompOperator::GtE,
        NotEq => CompOperator::NotEq,
        EqEq => CompOperator::Eq,
        _ => panic!(),
    };
    let span = left.span + right.span;
    Node::new(
        Expr::CompOperation {
            left: Box::new(left),
            op: Node::new(astop, op.span),
            right: Box::new(right),
        },
        span,
    )
}
