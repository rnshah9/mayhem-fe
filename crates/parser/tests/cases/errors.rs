use fe_common::db::TestDb;
use fe_common::diagnostics::diagnostics_string;
use fe_common::SourceFileId;
use fe_parser::grammar::{expressions, functions, module};
use fe_parser::Parser;
use insta::assert_snapshot;

pub fn err_string<F, T>(test_name: &str, mut parse_fn: F, src: &str) -> String
where
    F: FnMut(&mut Parser) -> T,
    T: std::fmt::Debug,
{
    let mut db = TestDb::default();
    let id = SourceFileId::new_local(&mut db, test_name, src.into());
    let mut parser = Parser::new(id, src);
    let _ = parse_fn(&mut parser);

    diagnostics_string(&db, &parser.diagnostics)
}

macro_rules! test_parse_err {
    ($name:ident, $parse_fn:expr, $src:expr) => {
        #[cfg(not(target_arch = "wasm32"))]
        #[test]
        fn $name() {
            assert_snapshot!(err_string(stringify!($name), $parse_fn, $src));
        }

        #[cfg(target_arch = "wasm32")]
        #[wasm_bindgen_test::wasm_bindgen_test]
        fn $name() {
            let actual = err_string(stringify!($name), $parse_fn, $src);
            let expected = include_str!(concat!(
                "snapshots/cases__errors__",
                stringify!($name),
                ".snap"
            ))
            .splitn(3, "---\n")
            .last()
            .unwrap();
            pretty_assertions::assert_eq!(actual.trim(), expected.trim());
        }
    };
}

// These tests use the insta crate. insta will automatically generate the
// snapshot file on the first run.

test_parse_err! { contract_invalid_version_requirement, module::parse_module, r#"
pragma 0.o
contract C {}"#
}

test_parse_err! { contract_missing_version_requirement, module::parse_module, r#"
pragma
contract C {}
"#
}

test_parse_err! { contract_bad_name, module::parse_module, "contract 1X {\n x: u8 \n}" }
test_parse_err! { contract_field_after_def, module::parse_module, r#"
contract C {
  fn f() {}
  x: u8
}"#
}

test_parse_err! { type_desc_path_number, module::parse_module, "type Foo = some::mod::Foo::5000" }
test_parse_err! { module_pub_event, module::parse_module, "pub event E{\n  x: u8}" }
test_parse_err! { contract_pub_event, module::parse_module, "contract C {\n pub event E{\n  x: u8 \n}}" }
test_parse_err! { contract_const_pub, module::parse_module, "contract C {\n const pub x: u8\n}" }
test_parse_err! { contract_const_fn, module::parse_module, "contract C {\n const fn f() {}\n}" }
test_parse_err! { emit_no_args, functions::parse_stmt, "emit x" }
test_parse_err! { emit_expr, functions::parse_stmt, "emit x + 1" }
test_parse_err! { emit_bad_call, functions::parse_stmt, "emit MyEvent(1)()" }
test_parse_err! { expr_bad_prefix, expressions::parse_expr, "*x + 1" }
test_parse_err! { expr_path_left, expressions::parse_expr, "(1 + 2)::foo::bar" }
test_parse_err! { expr_path_right, expressions::parse_expr, "foo::10::bar" }
test_parse_err! { expr_dotted_number, expressions::parse_expr, "3.14" }
test_parse_err! { expr_call_eq_label, expressions::parse_expr, "foo(bar=1, baz = 2)" }
test_parse_err! { expr_assignment, expressions::parse_expr, "1 + (x = y)" }
test_parse_err! { for_no_in, functions::parse_stmt, "for x {}" }
test_parse_err! { fn_no_args, module::parse_module, "fn f {\n  return 5\n}" }
test_parse_err! { fn_unsafe_pub, module::parse_module, "unsafe pub fn f() {\n  return 5 }" }
test_parse_err! { fn_def_kw, module::parse_module, "contract C {\n pub def f(x: u8){\n  return x \n}\n}" }

test_parse_err! { fn_invalid_bound, module::parse_module, "pub fn f<T:(u8, u8)>() {}" }
test_parse_err! { use_bad_name, module::parse_use, "use x as 123" }
test_parse_err! { module_bad_stmt, module::parse_module, "if x { y }" }
test_parse_err! { module_nonsense, module::parse_module, "))" }
test_parse_err! { struct_bad_field_name, module::parse_module, "struct f {\n pub event }" }
test_parse_err! { stmt_vardecl_attr, functions::parse_stmt, "f.s : u" }
test_parse_err! { stmt_vardecl_tuple, functions::parse_stmt, "(a, x+1) : u256" }
test_parse_err! { stmt_vardecl_tuple_empty, functions::parse_stmt, "(a, ()) : u256" }
test_parse_err! { stmt_vardecl_subscript, functions::parse_stmt, "a[1] : u256" }
test_parse_err! { stmt_vardecl_missing_type_annotation, functions::parse_stmt, "let x = 1" }
test_parse_err! { stmt_vardecl_missing_type_annotation_2, functions::parse_stmt, "let x" }
test_parse_err! { stmt_vardecl_missing_type_annotation_3, functions::parse_stmt, "let x:" }
test_parse_err! { stmt_vardecl_invalid_type_annotation, functions::parse_stmt, "let x: y + z" }
test_parse_err! { stmt_vardecl_invalid_name, functions::parse_stmt, "let x + y: u8" }
test_parse_err! { number_end_with_underscore, functions::parse_stmt, "42_42_"}
test_parse_err! { array_old_syntax, functions::parse_stmt, "let x: u8[10]" }
test_parse_err! { array_old_syntax_invalid, functions::parse_stmt, "let x: u8[10" }
test_parse_err! { self_const, module::parse_module, "const self: u8 = 10" }
test_parse_err! { self_contract, module::parse_module, "contract self {}" }
test_parse_err! { self_struct, module::parse_module, "struct self {}" }
test_parse_err! { self_fn, module::parse_module, "pub fn self() {}" }
test_parse_err! { self_use1, module::parse_module, "use self as bar" }
test_parse_err! { self_use2, module::parse_module, "use bar as self" }

// assert_snapshot! doesn't like the invalid escape code
#[test]
fn string_invalid_escape() {
    let err = err_string(
        "string_invalid_escape",
        expressions::parse_expr,
        r#""a string \c ""#,
    );
    assert_snapshot!(err);
}
