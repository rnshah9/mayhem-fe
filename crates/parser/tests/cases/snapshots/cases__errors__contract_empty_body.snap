---
source: crates/parser/tests/cases/errors.rs
expression: "err_string(stringify!(contract_empty_body), module::parse_module,\n           \"contract X {} \\ncontract Y {\\n  x: u8 \\n}\")"

---

